Robustify operations propagation

gterzian opened the pull request:
FIX https://github.com/massalabs/massa/issues/3054

gterzian commented:
@AurelienFT Could you please run the `charge.py` command with the simulator on this branch? I'm having some troubles installing it locally(see https://github.com/massalabs/massa-network-simulator/issues/35). Thanks!

gterzian commented:
@massalabs/core-team Ready for review.

Haven't been able to run the simulator yet(see the various [issues](https://github.com/massalabs/massa-network-simulator/issues)I filed), so if someone who has could run this on through it, it would be appreciated.

gterzian commented on massa-protocol-worker/src/protocol_worker.rs:
```diff
@@ -474,7 +490,22 @@ impl ProtocolWorker {
             }
             ProtocolCommand::PropagateOperations(storage) => {
-                self.propagate_ops(&storage).await;
+                // Note: should we claim refs locally?
```
Wondering about this.

damip replied:
That's an important question.

For now I think we can just ignore it because we will end up sending only the ops that are kept in the pool.
In the future (eg. when the pool gets separated out) we should somehow keep the refs to those ops for the duration of the propagation.

damip commented on massa-protocol-worker/src/protocol_worker.rs:
```diff
@@ -213,6 +216,7 @@ impl ProtocolWorker {
                 config.operation_batch_buffer_capacity,
             ),
             storage,
+            operations_to_announce: Default::default(),
```
you can preallocate it with the max op batch announcement size

damip commented on massa-protocol-worker/src/protocol_worker.rs:
```diff
@@ -292,9 +296,14 @@ impl ProtocolWorker {
 
-                // operation ask timer
+                // operation ask, and announce, timer
                 _ = &mut operation_batch_proc_period_timer => {
```
It should be a different timer

gterzian replied:
Why? It seems to make sense to announce, and ask, at the same interval.

gterzian replied:
Ah ok I see that the ask is not done at a fixed interval.

damip replied:
The operation asking system and the announcement batching system are separate, independent mechanisms. Let's keep the configs separate for more flexibility

damip commented on massa-protocol-worker/src/protocol_worker.rs:
```diff
@@ -344,6 +346,20 @@ impl ProtocolWorker {
+        // If the buffer is full,
+        // announce operations immediately,
+        // clearing the data at the same time.
+        if self.operations_to_announce.len() > self.config.max_known_ops_size {
```
the buffer size should not be max_known_ops_size but another specific config parameter (eg. op_announcement_buffer_size)

gterzian replied:
How about we re-use `operation_batch_buffer_capacity`?

damip replied:
The operation asking system and the announcement batching system are separate, independent mechanisms. Let's keep the configs separate for more flexibility

gterzian replied:
Ok, done.

damip commented on massa-protocol-worker/src/protocol_worker.rs:
```diff
@@ -344,6 +346,20 @@ impl ProtocolWorker {
         }
     }
```
at the end of announce_ops, reset the announcement timer

damip commented on massa-protocol-worker/src/protocol_worker.rs:
```diff
@@ -309,33 +318,26 @@ impl ProtocolWorker {
+    /// Side effects:
+    /// - notes nodes as knowing about those operations from now on.
+    /// - empties the buffer of operations to announce.
+    async fn announce_ops(&mut self) {
```
* reset the announcement timer
* if the buffer is empty, quit to avoid iterating on nodes

gterzian replied:
> reset the announcement timer

I think I'll just use a tokio interval.

damip replied:
@gterzian  the 1st point (resetting the timer) was not addressed in the case where the buffer is full and we send immediately

gterzian replied:
@damip Does it matter? We could also just remove that, or do it first and leave the newly added ops in the buffer for the next tick. We can tweak it later based on how the batching actually turns out I think.

damip replied:
* if we only send on timer, sometimes we will get huge batches and extra latency for no reason
* if we send on buffer full but only reset the timer when it ticks, sometimes we will send tiny batches for no reason except that we sent a full buffer just before
* if we reset the timer every time we send, we get the best compromise: low latency when the buffer is full, otherwise maximal buffer size at a given capped latency

damip commented:
@gterzian

Very good work ! (with a few caveats)

Some missing things to make it really effective:
* can you integrate https://github.com/massalabs/massa/pull/3095 in this PR now that we have batches ?
* once it's done, can you also try again with the simulator that was corrected by @sydhds  and @AurelienFT ?
* then finish the robustify checklist in https://github.com/massalabs/massa/issues/3054

gterzian commented:
@damip Thanks for the review, comments addressed. I suggest we merge this one first, and then I can integrate the parallelization via a rebase of [that branch](https://github.com/massalabs/massa/pull/3095).

damip commented on massa-node/base_config/config.toml:
```diff
@@ -85,6 +85,8 @@
     # All operations asked are prune each `operation_asked_pruning_period` millisecond
     asked_operations_pruning_period = 100000
+    # Interval at which operations are announced in batches.
+    operation_announcement_interval = 500
```
500ms is very high. Let's set it to 150ms

damip commented on massa-protocol-exports/src/tests/tools.rs:
```diff
@@ -274,6 +274,7 @@ pub fn create_protocol_config() -> ProtocolConfig {
         operation_batch_buffer_capacity: 1000,
         operation_batch_proc_period: 200.into(),
         asked_operations_pruning_period: 500.into(),
+        operation_announcement_interval: 500.into(),
```
150 ms

AurelienFT commented on massa-protocol-worker/src/protocol_worker.rs:
```diff
@@ -292,9 +303,24 @@ impl ProtocolWorker {
+                    operation_announcement_interval.set(sleep_until(next_tick));
+                }
+
+                // operation ask, and announce, timer
```
The documentation seems old here

gterzian replied:
yes, removed.

gterzian commented:
@AurelienFT @damip Thanks for the reviews, I think I have addressed all your comments. The protocol worker test suite passes, however I still haven't been able to run it on the simulator, and it doesn't look like I'll be able to for today.

damip commented:
Ok, @sydhds  can you take over for us to finish in time ?

damip commented:
@gterzian  It's crucial that you figure out how to run the simulator, we can't go without it anymore. Let's make sure you have it for tomorrow, and open issues if anything goes wrong to get help until it works.
